src/collapse/perf.rs

```diff
@@ -324,6 +387,91 @@ impl PerfState {
         }
         Ok(())
     }
+
+    // Use addr2line to determine the symbol to use for each program counter address
+    // (as opposed to using the symbol names that perf script produces)
+    fn un_inline(&mut self, addr: &str, module: &str) -> result::Result<String, Error> {
+        let addr = u64::from_str_radix(addr, 16).map_err(|e| {
+            Error::Parse(format!(
+                "unable to parse {} as hex address, caused by: {}",
+                addr, e
+            ))
+        })?;
+
+        if let Entry::Vacant(entry) = self.addr2line_contexts.entry(module.into()) {
+            let bytes = fs::read(module)?;
```

Comment:
Sounds good to me. I just wanted to avoid unsafe.

Reply:
Hmm, that's a good point, though I suspect the performance speed-up here is worth it. It's not immediately clear to me why `Mmap::map` is `unsafe` in the first place, and the docs don't specify why, but then I found https://github.com/danburkert/memmap-rs/pull/66. It seems like the concern is that the underlying file may be modified, which would cause undefined behavior :/ I still think this is the right thing to do, as we're not likely to run for very long, but perhaps document that just above the `unsafe` line?

Reply:
Agreed. It's worth it, especially for larger files.

Reply:
In particular because `addr2line` only needs the debug symbols, not the rest of the binary. With `fs::read`, we'll end up reading in the entire file!